src/walker.rs

```diff
@@ -0,0 +1,246 @@
+            WalkOrder::Preorder => &mut self.pre_callbacks,
+            WalkOrder::Postorder => &mut self.post_callbacks,
+        }
+    }
+
+    pub fn visit<O, F: 'a + FnMut(Node, O, T) -> Result<T, E>>(
+        &mut self,
+        walk_order: WalkOrder,
+        mut f: F,
+    ) -> &mut Self
+    where
+        OpType: TryInto<O>,
+    {
+        let cb = move |n, o: OpType, t| match o.try_into() {
+            Ok(x) => f(n, x, t),
+            _ => Ok(t),
+        };
+        self.mut_callbacks(walk_order).push(cb.into());
+        self
+    }
+
+    pub fn walk(&mut self, mut t: T) -> Result<T, E> {
+        // We intentionally avoid recursion so that we can robustly accept very deep hugrs
+        let mut worklist = vec![WorkItem::Visit(self.hugr.root())];
+
+        while let Some(wi) = worklist.pop() {
+            match wi {
+                WorkItem::Visit(n) => {
+                    worklist.push(WorkItem::Callback(WalkOrder::Postorder, n));
+                    let mut pushed_children = HashSet::new();
+                    // We intend to only visit direct children.
+                    //
+                    // If the nodes children form a dataflow sibling graph we
+                    // visit them in post dfs order. This traversal is not
+                    // guaranteed to visit all nodes, only those reachable from
+                    // the input node. (For example, LoadConstant nodes may not
+                    // be reachable from the input node). So we do a second
+                    // unordered traversal afterwards.
+                    if let Some([input, _]) = self.hugr.get_io(n) {
+                        use crate::hugr::views::PetgraphWrapper;
+                        let wrapper = PetgraphWrapper { hugr: self.hugr };
+                        let mut dfs = ::petgraph::visit::DfsPostOrder::new(&wrapper, input);
+                        while let Some(x) = dfs.next(&wrapper) {
+                            worklist.push(WorkItem::Visit(x));
+                            pushed_children.insert(x);
+                        }
+                    }
+
+                    let rest_children = self
+                        .hugr
+                        .children(n)
+                        .filter(|x| !pushed_children.contains(x))
+                        .collect_vec();
+                    // extend in reverse so that the first child is the top of the stack
+                    worklist.extend(rest_children.into_iter().rev().map(WorkItem::Visit));
+                    worklist.push(WorkItem::Callback(WalkOrder::Preorder, n));
+                }
+                WorkItem::Callback(order, n) => {
+                    let optype = self.hugr.get_optype(n);
+                    for cb in self.mut_callbacks(order).iter_mut() {
+                        // this clone is unfortunate, to avoid this we would
+                        // need a TryInto variant like:
+                        // try_into(&O) -> Option<&T>
```

Comment:
I bet there is some way with `enum_dispatch` but I'll leave that to someone Rustier than me.... @aborgna-q ? ;)

A clone is no big deal, I'm happy with that